src/main/java/io/confluent/connect/elasticsearch/OffsetTracker.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.connect.elasticsearch;
+
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.ErrantRecordReporter;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+import static java.util.stream.Collectors.toMap;
+
+class OffsetTracker {
+
+  private static final Logger log = LoggerFactory.getLogger(OffsetTracker.class);
+
+  static class OffsetState {
+
+    private static final Supplier<Boolean> TRUE = () -> true;
+    private final long offset;
+    private final AtomicReference<Supplier<Boolean>> processed = new AtomicReference<>();
+
+    OffsetState(long offset) {
+      this.offset = offset;
+    }
+
+    /**
+     * Marks the offset as processed (ready to report to preCommit), if it has not been marked yet
+     */
+    public void markProcessed() {
+      markProcessed(TRUE);
+    }
+
+    /**
+     * Provides a supplier that will determine if the record was processed or not.
+     * It does nothing is markProcessed was called before.
+     * Intended for async operations that return a Future and don't provide a callback we can
+     * hook to. (e.g. {@link ErrantRecordReporter#report(SinkRecord, Throwable)})
+     */
+    public void markProcessed(Supplier<Boolean> processedSupplier) {
```

Comment:
I am not a fan of this interface. It seems overly complex to have `markProcessed` either mark the state or provide a guarantee to mark the state some time in the future. Essentially, it is trying to do the job of marking state + Futures. Since it seems to exist for the sole purpose of `ErrantRecordReporter` , would it be easier to wrap that `Future` in a `CompletableFuture` and chain the call to mark state after it has completed?

Reply:
Thanks Nigel, I thought about that but the main problem is that the only way to wrap a `Future` in a `CompletableFuture` you either wait for the future in a separate thread, which in this case is not reasonable (this is per record) or you need to create a [separate event loop to schedule the result checking](http://www.thedevpiece.com/converting-old-java-future-to-completablefuture/), which also adds to the complexity.